aws/src/main/java/org/apache/iceberg/aws/LakeFormationAwsClientFactory.java

```diff
@@ -0,0 +1,153 @@
+  }
+
+  @Override
+  public S3Client s3() {
+    if (isRegisteredWithLakeFormation()) {
+      S3ClientBuilder builder = S3Client.builder()
+          .endpointOverride(URI.create(s3Endpoint()))
+          .credentialsProvider(new LakeFormationCredentialsProvider(getLakeFormationClient(), buildTableArn()))
+          .region(Region.of(region()))
+          .httpClientBuilder(UrlConnectionHttpClient.builder());
+      return builder.build();
+    } else {
+      return super.s3();
+    }
+  }
+
+  @Override
+  public KmsClient kms() {
+    if (isRegisteredWithLakeFormation()) {
+      KmsClientBuilder builder = KmsClient.builder()
+          .credentialsProvider(new LakeFormationCredentialsProvider(getLakeFormationClient(), buildTableArn()))
+          .region(Region.of(region()))
+          .httpClientBuilder(UrlConnectionHttpClient.builder());
+      return builder.build();
+    } else {
+      return super.kms();
+    }
+  }
+
+  private boolean isRegisteredWithLakeFormation() {
+    Preconditions.checkArgument(dbName != null && !dbName.isEmpty(), "db name can not be empty");
+    Preconditions.checkArgument(tableName != null && !tableName.isEmpty(), "table name can not be empty");
+
+    GetTableResponse response = glue().getTable(GetTableRequest.builder()
+        .catalogId(glueCatalogId)
+        .databaseName(dbName)
+        .name(tableName)
+        .build());
+    return response.table().isRegisteredWithLakeFormation();
+  }
+
+  private String buildTableArn() {
+    Preconditions.checkArgument(region != null && !region.isEmpty(), "region can not be empty");
+    Preconditions.checkArgument(glueAccountId != null && !glueAccountId.isEmpty(), "glueAccountId can not be empty");
+    Preconditions.checkArgument(dbName != null && !dbName.isEmpty(), "db name can not be empty");
+    Preconditions.checkArgument(tableName != null && !tableName.isEmpty(), "table name can not be empty");
+    String partitionName = PartitionMetadata.of(Region.of(region)).id();
+    return String.format("arn:%s:glue:%s:%s:table/%s/%s",
+        partitionName,
+        region,
+        glueAccountId,
+        dbName,
+        tableName);
+  }
+
+  private LakeFormationClient getLakeFormationClient() {
+    return LakeFormationClient.builder()
+        .applyMutation(this::configure)
+        .build();
+  }
+
+  static class LakeFormationCredentialsProvider implements AwsCredentialsProvider {
+    private LakeFormationClient client;
+    private String tableArn;
+
+    LakeFormationCredentialsProvider(LakeFormationClient lakeFormationClient, String tableArn) {
+      this.client = lakeFormationClient;
+      this.tableArn = tableArn;
+    }
+
+    @Override
+    public AwsCredentials resolveCredentials() {
+      GetTemporaryGlueTableCredentialsRequest getTemporaryGlueTableCredentialsRequest =
+          GetTemporaryGlueTableCredentialsRequest.builder()
+              .tableArn(tableArn)
+              .permissions(Permission.ALL)
```

Comment:
add an inline comment that "Iceberg cannot distinguish caller's intended operation, always request all possible permissions"

Reply:
done. thanks!